File: opensource/test/src/main/java/com/sos/scheduler/engine/test/TestSchedulerController.java
package com.sos.scheduler.engine.test;

import com.google.common.base.Predicate;
import com.google.common.base.Splitter;
import com.google.common.collect.ImmutableMap;
import com.sos.scheduler.engine.common.system.Files;
import com.sos.scheduler.engine.common.time.Time;
import com.sos.scheduler.engine.data.log.ErrorLogEvent;
import com.sos.scheduler.engine.data.log.SchedulerLogLevel;
import com.sos.scheduler.engine.eventbus.*;
import com.sos.scheduler.engine.kernel.Scheduler;
import com.sos.scheduler.engine.kernel.log.PrefixLog;
import com.sos.scheduler.engine.kernel.settings.SettingName;
import com.sos.scheduler.engine.kernel.settings.Settings;
import com.sos.scheduler.engine.kernel.util.Hostware;
import com.sos.scheduler.engine.kernel.util.ResourcePath;
import com.sos.scheduler.engine.main.CppBinaries;
import com.sos.scheduler.engine.main.CppBinary;
import com.sos.scheduler.engine.main.SchedulerState;
import com.sos.scheduler.engine.test.binary.CppBinariesDebugMode;
import com.sos.scheduler.engine.test.binary.TestCppBinaries;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import javax.annotation.Nullable;
import java.io.File;
import java.util.ArrayList;
import java.util.List;
import java.util.concurrent.atomic.AtomicInteger;

import static com.google.common.base.Preconditions.checkState;
import static com.google.common.base.Strings.nullToEmpty;
import static com.google.common.collect.Iterables.concat;
import static com.google.common.collect.Iterables.toArray;
import static com.sos.scheduler.engine.common.system.Files.makeTemporaryDirectory;
import static com.sos.scheduler.engine.common.system.Files.tryRemoveDirectoryRecursivly;
import static java.util.Arrays.asList;

public class TestSchedulerController extends DelegatingSchedulerController implements EventHandlerAnnotated {
    private static final Logger logger = LoggerFactory.getLogger(TestSchedulerController.class);
    private static final String workDirectoryPropertyName = "com.sos.scheduler.engine.test.directory";
    public static final Time shortTimeout = Time.of(15);
    private static final AtomicInteger instanceCount = new AtomicInteger(0);

    private final List<Runnable> closingRunnables = new ArrayList<Runnable>();
    private final SchedulerEventBus eventBus = getEventBus();
    private final Thread thread = Thread.currentThread();
    private final Environment environment;
    private final Predicate<ErrorLogEvent> expectedErrorLogEventPredicate;
    private final CppBinariesDebugMode debugMode;

    private boolean terminateOnError = true;
    private boolean isPrepared = false;
    private String logCategories = "";
    private Scheduler _scheduler = null;   // Unterstrich, damit IntelliJ-Scala-Plugin scheduler() findet, Zschimmer 9.12.2011

    public TestSchedulerController(Class<?> testClass, ResourcePath configurationResourcePath,
            @Nullable ImmutableMap<String,String> nameMap,
            @Nullable ResourceToFileTransformer fileTransformer,
            Predicate<ErrorLogEvent> expectedErrorLogEventPredicate,
            CppBinariesDebugMode debugMode) {
        super(testClass.getName());
        logger.debug(testClass.getName());
        instanceCount.addAndGet(1);
        environment = new Environment(configurationResourcePath, workDirectory(testClass), nameMap, fileTransformer);
        this.expectedErrorLogEventPredicate = expectedErrorLogEventPredicate;
        this.debugMode = debugMode;
        setSettings(Settings.of(SettingName.jobJavaClasspath, System.getProperty("java.class.path")));
    }

    private File workDirectory(Class<?> testClass) {
        String p = System.getProperty(workDirectoryPropertyName);
        if (p != null) {
            new File(p).mkdir();    // Das Unterverzeichnis failsafe-reports wird verzögert angelegt. Dem kommen wir zuvor.
            File result = new File(p, testClass.getName());
            makeCleanDirectory(result);
            return result;
        } else {
            final File result = makeTemporaryDirectory();
            closingRunnables.add(new Runnable() {
                @Override public void run() { tryRemoveDirectoryRecursivly(result); }
            });
            return result;
        }
    }

    private static void makeCleanDirectory(File directory) {
        //Files.makeDirectory(directory.getParentFile());
        Files.makeDirectory(directory);
        Files.removeDirectoryContentRecursivly(directory);
    }

    @Override public final void close() {
        try {
            getDelegate().close();
        } finally {
            for (Runnable r: closingRunnables) r.run();
        }
    }

    /** Bricht den Test mit Fehler ab, wenn ein {@link com.sos.scheduler.engine.data.log.ErrorLogEvent} ausgelöst worden ist. */
    public final void setTerminateOnError(boolean o) {
        getDelegate().checkIsNotStarted();
        terminateOnError = o;
    }

    public final void setLogCategories(String o) {
        getDelegate().checkIsNotStarted();
        logCategories = o;
    }

    /** @param timeout Wenn ab Bereitschaft des Schedulers mehr Zeit vergeht, wird eine Exception ausgelöst */
    public final void runScheduler(Time timeout, String... args) {
        activateScheduler(args);
        waitForTermination(timeout);
    }

    /** Startet den Scheduler und wartet, bis er aktiv ist. */
    public final void activateScheduler(String... args) {
        startScheduler(args);
        waitUntilSchedulerIsActive();
    }

    @Override public final void startScheduler(String... args) {
        prepare();
        String extraOptions = nullToEmpty(System.getProperty(TestSchedulerController.class.getName() +".options"));
        Iterable<String> allArgs = concat(
                environment.standardArgs(cppBinaries(), logCategories),
                Splitter.on(",").omitEmptyStrings().split(extraOptions),
                asList(args));
        getDelegate().startScheduler(toArray(allArgs, String.class));
    }

    public void prepare() {
        if (!isPrepared) {
            registerEventHandler(this);
            environment.prepare();
            getDelegate().loadModule(cppBinaries().file(CppBinary.moduleFilename));
            isPrepared = true;
        }
    }

    public final Scheduler scheduler() {
        if (_scheduler == null) {
            automaticStart();
            waitUntilSchedulerIsActive();
        }
        return _scheduler;
    }

    /** Wartet, bis das Objekt {@link Scheduler} verfügbar ist. */
    public final void waitUntilSchedulerIsActive() {
        Scheduler previous = _scheduler;
        _scheduler = getDelegate().waitUntilSchedulerState(SchedulerState.active);
        if (_scheduler == null) throw new RuntimeException("Scheduler aborted before startup");
        if (previous == null && terminateOnError) checkForErrorLogLine();
    }

    public final void waitForTermination(Time timeout) {
        boolean ok = tryWaitForTermination(timeout);
        if (!ok) throw new SchedulerRunningAfterTimeoutException(timeout);
    }

    private void automaticStart() {
        if (!getDelegate().isStarted()) {
            checkState(Thread.currentThread() == thread, "TestSchedulerController.automaticStart() must be called in constructing thread");
            throw new RuntimeException("JobScheduler is not active yet");
//            startScheduler();
        }
    }

    private void checkForErrorLogLine() {
        String lastErrorLine = _scheduler.injector().getInstance(PrefixLog.class).lastByLevel(SchedulerLogLevel.error);
        if (!lastErrorLine.isEmpty())
            throw new RuntimeException("Test terminated after error log line: "+ lastErrorLine);
    }

    @EventHandler public final void handleEvent(ErrorLogEvent e) {
        if (!expectedErrorLogEventPredicate.apply(e) && terminateOnError)
            terminateAfterException(new RuntimeException("Test terminated after error log line: "+ e.getLine()));
    }

    @EventHandler @HotEventHandler  // Beide, weil das EventHandlerFailedEvent wird nur innerhalb von Hot- oder ColdEventBus veröffentlicht wird.
    public final void handleEvent(EventHandlerFailedEvent e) {
        if (terminateOnError)
            terminateAfterException(e.getThrowable());
    }

    /** Eine Exception in {@code runnable} beendet den Scheduler. */
    public final Thread newThread(Runnable runnable) {
        ThreadTerminationHandler h = new ThreadTerminationHandler() {
            @Override public void onThreadTermination(@Nullable Throwable t) {
                if (t != null) terminateAfterException(t);
            }
        };
        return new Thread(new TestThreadRunnable(runnable, h), runnable.toString());
    }

    /** Rechtzeitig aufrufen, dass kein Event verloren geht. */
    public final EventPipe newEventPipe() {
        EventPipe result = new EventPipe(shortTimeout);
        registerEventHandler(result);
        return result;
    }

    private void registerEventHandler(final EventHandlerAnnotated o) {
        eventBus.registerAnnotated(o);
        closingRunnables.add(new Runnable() {
            @Override public void run() { eventBus.unregisterAnnotated(o); }
        });
    }

    public final boolean isStarted() {
        return getDelegate().isStarted();
    }

    public final void useDatabase() {
        String dbName = Hostware.h2DatabasePath("mem:scheduler-"+instanceCount);
        setSettings(Settings.of(SettingName.dbName, dbName));
    }

    public final Environment environment() {
        return environment;
    }

    public final CppBinaries cppBinaries() {
        return TestCppBinaries.cppBinaries(debugMode);
    }

    /** @param testClass Test-Klasse, für Benennung des Scheduler-Arbeitsverzeichnisses und Ort der Konfigurationsresourcen. */
    public static TestSchedulerController of(Class<?> testClass) {
        return builder(testClass).build();
    }

    /** @param testClass Test-Klasse, für Benennung des Scheduler-Arbeitsverzeichnisses und Ort der Konfigurationsresourcen. */
    public static TestSchedulerControllerBuilder builder(Class<?> testClass) {
        return new TestSchedulerControllerBuilder(testClass);
    }
}

File: opensource/test/src/main/scala/com/sos/scheduler/engine/test/TestSchedulerController.scala
package com.sos.scheduler.engine.test

import TestSchedulerController._
import _root_.scala.Some
import _root_.scala.collection.JavaConversions._
import _root_.scala.reflect.ClassTag
import _root_.scala.sys.error
import com.google.common.base.Splitter
import com.google.common.base.Strings.nullToEmpty
import com.google.common.base.Throwables._
import com.sos.scheduler.engine.common.scalautil.{HasCloser, Logger}
import com.sos.scheduler.engine.common.time.ScalaJoda._
import com.sos.scheduler.engine.common.xml.XmlUtils.{loadXml, prettyXml}
import com.sos.scheduler.engine.data.log.ErrorLogEvent
import com.sos.scheduler.engine.data.log.SchedulerLogLevel
import com.sos.scheduler.engine.eventbus._
import com.sos.scheduler.engine.kernel.Scheduler
import com.sos.scheduler.engine.kernel.log.PrefixLog
import com.sos.scheduler.engine.kernel.scheduler.HasInjector
import com.sos.scheduler.engine.kernel.settings.{CppSettingName, CppSettings}
import com.sos.scheduler.engine.kernel.util.Hostware
import com.sos.scheduler.engine.main.{SchedulerThreadController, CppBinaries, CppBinary, SchedulerState}
import com.sos.scheduler.engine.test.binary.{CppBinariesDebugMode, TestCppBinaries}
import com.sos.scheduler.engine.test.configuration.{HostwareDatabaseConfiguration, JdbcDatabaseConfiguration, TestConfiguration}
import com.sos.scheduler.engine.test.scala.SchedulerTestImplicits._
import java.io.File
import java.sql.{Connection, DriverManager}
import org.joda.time.Duration

abstract class TestSchedulerController
extends DelegatingSchedulerController
with HasCloser
with HasInjector
with EventHandlerAnnotated {

  def testConfiguration: TestConfiguration

  def environment: TestEnvironment

  private val testName = testConfiguration.testClass.getName
  protected final lazy val delegate = new SchedulerThreadController(testName, cppSettings(testName, testConfiguration, environment.databaseDirectory))
  private val eventBus: SchedulerEventBus = getEventBus
  private val thread = Thread.currentThread
  private val debugMode = testConfiguration.binariesDebugMode getOrElse CppBinariesDebugMode.debug
  private val logCategories = testConfiguration.logCategories + " " + sys.props.getOrElse("scheduler.logCategories", "").trim

  private var isPrepared: Boolean = false
  private var _scheduler: Scheduler = null
  @volatile private var suppressTerminateOnError = false

  private val jdbcUrlOption: Option[String] =
    testConfiguration.database collect {
      case c: JdbcDatabaseConfiguration =>
        Class forName c.jdbcClassName
        c.testJdbcUrl(testName, environment.databaseDirectory)
    }

  override def close() {
    try delegate.close()
    super.close()
  }

  /** Startet den Scheduler und wartet, bis er aktiv ist. */
  def activateScheduler(args: java.lang.Iterable[String]) {
    startScheduler(iterableAsScalaIterable(args).toSeq: _*)
    waitUntilSchedulerIsActive()
  }

  /** Startet den Scheduler und wartet, bis er aktiv ist. */
  def activateScheduler() {
    activateScheduler(Nil: _*)
  }

  /** Startet den Scheduler und wartet, bis er aktiv ist. */
  def activateScheduler(args: String*) {
    startScheduler(args: _*)
    waitUntilSchedulerIsActive()
  }

  def startScheduler(args: java.lang.Iterable[String]) {
    startScheduler(iterableAsScalaIterable(args).toSeq)
  }

  def startScheduler(args: String*) {
    prepare()
    val extraOptions = nullToEmpty(System.getProperty(classOf[TestSchedulerController].getName + ".options"))
    val allArgs = Seq() ++
        environment.standardArgs(cppBinaries, logCategories) ++
        Splitter.on(",").omitEmptyStrings.split(extraOptions) ++
        testConfiguration.mainArguments ++
        args
    delegate.startScheduler(allArgs)
  }

  def prepare() {
    if (!isPrepared) {
      registerEventHandler(this)
      environment.prepare()
      delegate.loadModule(cppBinaries.file(CppBinary.moduleFilename))
      isPrepared = true
    }
  }

  def scheduler: Scheduler = {
    if (_scheduler == null) {
      automaticStart()
      waitUntilSchedulerIsActive()
    }
    _scheduler
  }

  /** Wartet, bis das Objekt [[com.sos.scheduler.engine.kernel.Scheduler]] verfügbar ist. */
  def waitUntilSchedulerIsActive() {
    val previous = _scheduler
    _scheduler = delegate.waitUntilSchedulerState(SchedulerState.active)
    if (_scheduler == null) throw new RuntimeException("Scheduler aborted before startup")
    if (previous == null && testConfiguration.terminateOnError) checkForErrorLogLine()
  }

  def waitForTermination(timeout: Duration) {
    val ok = tryWaitForTermination(timeout)
    if (!ok) {
      val x = new SchedulerRunningAfterTimeoutException(timeout)
      logger warn x.toString
      val cmd = "<show_state what='folders jobs job_params job_commands tasks task_queue job_chains orders remote_schedulers operations'/>"
      logger warn cmd
      logger warn prettyXml(loadXml(scheduler.uncheckedExecuteXml(cmd)))
      throw x
    }
  }

  private def automaticStart() {
    if (!delegate.isStarted) {
      if (Thread.currentThread ne thread)  throw new IllegalStateException("TestSchedulerController.automaticStart() must be called in constructing thread")
      error("JobScheduler is not active yet")
    }
  }

  private def checkForErrorLogLine() {
    val lastErrorLine = _scheduler.instance[PrefixLog].lastByLevel(SchedulerLogLevel.error)
    if (!lastErrorLine.isEmpty) error("Test terminated after error log line: " + lastErrorLine)
  }

  def suppressingTerminateOnError[A](f: => A): A = {
    require(!suppressTerminateOnError)
    suppressTerminateOnError = true
    try {
      val result = f
      getEventBus.dispatchEvents()   // Damit handleEvent(ErrorLogEvent) wirklich jetzt gerufen wird, solange noch suppressTerminateOnError gilt
      result
    }
    finally suppressTerminateOnError = false
  }

  @EventHandler
  def handleEvent(e: ErrorLogEvent) {
    if (testConfiguration.terminateOnError && !suppressTerminateOnError && !testConfiguration.ignoreError(e.getCodeOrNull) && !testConfiguration.errorLogEventIsExpected(e))
      terminateAfterException(error(s"Test terminated after error log line: ${e.getLine}"))
  }

  @EventHandler @HotEventHandler
  def handleEvent(e: EventHandlerFailedEvent) {
    if (testConfiguration.terminateOnError) {
      logger.debug("SchedulerTest is aborted due to 'terminateOnError' and error: " + e)
      terminateAfterException(e.getThrowable)
    }
  }

  final def instance[A](implicit c: ClassTag[A]): A =
    injector.getInstance(c.runtimeClass.asInstanceOf[Class[A]])

  final def injector =
    scheduler.injector

  /** Eine Exception in runnable beendet den Scheduler. */
  def newThread(runnable: Runnable) =
    new Thread {
      override def run() {
        try runnable.run()
        catch {
          case t: Throwable =>
            terminateAfterException(t)
            throw propagate(t)
        }
      }
    }

  /** Rechtzeitig aufrufen, dass kein Event verloren geht. */
  def newEventPipe(): EventPipe = {
    val result = new EventPipe(eventBus, shortTimeout.toDuration)
    registerEventHandler(result)
    result
  }

  private def registerEventHandler(o: EventHandlerAnnotated) {
    eventBus registerAnnotated o
    onClose { eventBus unregisterAnnotated o }
  }

  def isStarted =
    delegate.isStarted

  def cppBinaries: CppBinaries =
    TestCppBinaries.cppBinaries(debugMode)

  def newJDBCConnection(): Connection =
    DriverManager.getConnection(jdbcUrlOption.get)
}


object TestSchedulerController {
  final val shortTimeout = 15.s
  private val logger = Logger(getClass)

  def apply(testConfiguration: TestConfiguration): TestSchedulerController = {
    val _testConfiguration = testConfiguration
    new TestSchedulerController with ProvidesTestEnvironment {
      override lazy val testConfiguration = _testConfiguration
      lazy val environment = testEnvironment
    }
  }

  def apply(testConfiguration: TestConfiguration, testEnvironment: TestEnvironment): TestSchedulerController = {
    val _configuration = testConfiguration
    new TestSchedulerController {
      override lazy val testConfiguration = _configuration
      lazy val environment = testEnvironment
    }
  }

  private def cppSettings(testName: String, configuration: TestConfiguration, databaseDirectory: File): CppSettings =
    CppSettings(
      configuration.cppSettings
        + (CppSettingName.jobJavaClasspath -> System.getProperty("java.class.path"))
        ++ dbNameCppSetting(testName, configuration, databaseDirectory))

  private def dbNameCppSetting(testName: String, configuration: TestConfiguration, databaseDirectory: File): Option[(CppSettingName, String)] =
    configuration.database match {
      case Some(c: JdbcDatabaseConfiguration) =>
        Some(CppSettingName.dbName -> Hostware.databasePath(c.jdbcClassName, c.testJdbcUrl(testName, databaseDirectory)))
      case Some(c: HostwareDatabaseConfiguration) =>
        Some(CppSettingName.dbName -> c.hostwareString)
      case None =>
        None
    }
}
